### PR TITLE
Disable setting keys in the same ruleset with the same bindings

### PR DESCRIPTION
**Pull request by nanashi-1**
- Addresses #20107 

Can no longer set keys in the same ruleset with the same bindings but it may need feedback like a toast or an animation.

### CODE REVIEW

File: osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs [hunk 1]
```diff
@@ -530,6 +534,9 @@ public void UpdateKeyCombination(KeyCombination newCombination)
                 if (KeyBinding.RulesetName != null && !RealmKeyBindingStore.CheckValidForGameplay(newCombination))
                     return;
 
+                if (RulesetBindings.Select(k => k.KeyCombination).Contains(newCombination))
+                    return;
```
Comment (frenzibyte) [inline osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs:538]: So when a user inputs the same key in another action, it'll just silently ignore it? There should be visual feedback, at least flashing red on the key binding button.

File: osu.Game/Overlays/Settings/Sections/Input/KeyBindingsSubsection.cs [hunk 2]
```diff
@@ -49,7 +49,8 @@ private void load(RealmAccess realm)
                 Add(new KeyBindingRow(defaultGroup.Key, bindings.Where(b => b.ActionInt.Equals(intKey)).ToList())
                 {
                     AllowMainMouseButtons = Ruleset != null,
-                    Defaults = defaultGroup.Select(d => d.KeyCombination)
+                    Defaults = defaultGroup.Select(d => d.KeyCombination),
+                    RulesetBindings = bindings
```
Comment (frenzibyte) [inline osu.Game/Overlays/Settings/Sections/Input/KeyBindingsSubsection.cs]: Are you sure this works when changing a key binding after the subsection was loaded, and *then* attempting to use the same key in another action?

**Review by frenzibyte — COMMENTED** [review]
Passing the ruleset bindings all the way down to each `KeyButton` in order to handle against duplicate keys sounds like an awkward flow.

I haven't checked deeply how this goes performance-wise but you should be able to at least query the realm keybindings at the point of updating key combination, rather than passing a list on construction.

File: osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs (hunk 1 above)
Comment (nanashi-1) [reply osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs:538]: Is flashing red enough?

Comment (nanashi-1) [reply osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs:538]: Is the color `Color4.Red` enough?

Comment (frenzibyte) [reply osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs:538]: Sounds fine.

Comment (nanashi-1) [reply osu.Game/Overlays/Settings/Sections/Input/KeyBindingRow.cs:538]: It looks something like this with `Color4.Red` as the color of the box.

Screenshot from 2022-09-04 20-42-30

**Comment by peppy** [comment]
Keep in mind that this isn't enforcing anything. If the key bindings are already set they are still going to be usable.

I believe you'll also need some extra logic around this point:

https://github.com/ppy/osu/blob/31a447fda0532fcf15f8c8251ca890c533386492/osu.Game/Rulesets/UI/RulesetInputManager.cs#L205

**Review by peppy — CHANGES_REQUESTED** [review]
As proposed

**Comment by nanashi-1** [comment]
> Keep in mind that this isn't enforcing anything. If the key bindings are already set they are still going to be usable.
> 
> I believe you'll also need some extra logic around this point:
> 
> https://github.com/ppy/osu/blob/31a447fda0532fcf15f8c8251ca890c533386492/osu.Game/Rulesets/UI/RulesetInputManager.cs#L205

I don't quite understand what this does.

**Comment by peppy** [comment]
Let me explain another way: Your change is only blocking UI level changes. If the user already has bindings set in the realm database, or manually sets it externally, the changes you've made are not going to have any effect.

**Comment by peppy** [comment]
No test coverage was added to check the recent fix to the issue I mentioned. Should be added to make review of this PR easier.

**Pull request closed without merging**